### PR TITLE
bugfix - duplicate luadoc entries

### PR DESCRIPTION
**Pull request by johann1764**
**Bug Fix** / **New Feature**

This PR addresses the bug/feature described in #2829.

Baware that sorting takes time, even sorting the list of `@luafunc` occurences. Might take up to a second on your machine, or maybe even more.

## Summary
 - [x] luafunc_dup.sh
 - [x] nlua_pilot.c
 - [x] other nlua_*.c files
 - [x] add it to precommit
 - [x] handle multiple files
 - [x] simplify and improve the script with uniq

```shell
src/nlua_asteroid.c:
      2  * @luafunc setTimer
      2  * @luafunc state
src/nlua_faction.c:
      2  * @luafunc hit
src/nlua_jump.c:
      2  * @luafunc dest
src/nlua_naev.c:
      2  * @luafunc pause
src/nlua_outfit.c:
      2  * @luafunc get
      2  * @luafunc slot
src/nlua_pilot.c:
      2  * @luafunc colour
      2  * @luafunc setHealth
      2  * @luafunc setHilight
      2  * @luafunc setInvisible
      2  * @luafunc weapsetAdd
src/nlua_player.c:
      3  * @luafunc inventoryAdd
      2  * @luafunc tryBoard
src/nlua_shader.c:
      2  * @luafunc send
src/nlua_spfx.c:
      2  * @luafunc vel
src/nlua_spob.c:
      2  * @luafunc services
src/nlua_system.c:
      2  * @luafunc hidden
src/nlua_transform.c:
      2  * @luafunc translate
```

### CODE REVIEW

**Comment by bobbens** [comment]
> **Bug Fix** / **New Feature**
> 
> This PR addresses the bug/feature described in #2829.

Thanks!

> Baware that sorting takes time, even sorting the list of `@luafunc` occurences. Might take up to a second on your machine, or maybe even more.

Shouldn't be a problem. We can make it only run on `nlua_` files.

> ## Summary
> 
>     * [x]  luafunc_dup.sh
> 
>     * [x]  nlua_pilot.c
> 
>     * [x]  others
> 
>     * [x]  add it to precommit
> 
>     * [ ]  check this does need to handle multiple files (would be easy to fix)

Yes, I think it can be batched, so it should support something like `./script FILE1 FILE2 FILE3`, although they should be handled independently (no concatenation) as different files may have same named functions.

**Comment by johann1764** [comment]
> > Baware that sorting takes time, even sorting the list of `@luafunc` occurences. Might take up to a second on your machine, or maybe even more.
> 
> Shouldn't be a problem. We can make it only run on `nlua_` files.

Bad news: already took that into account. You should probably try to see whether this is too costly or not.

> Yes, I think it can be batched, so it should support something like `./script FILE1 FILE2 FILE3`, although they should be handled independently (no concatenation) as different files may have same named functions.

Now it can.

Even, it is better to do it all at the same time, because the script waits until the end before exiting with the code, meaning you will have all files scanned with their redundancies before the test fails, which is nicer to have it fail after one failed file, and having to repeat until no new file fails. Not sure I am clear, here.

**Comment by johann1764** [comment]
If we really need the test to be faster, we can only check successive duplicates, which covers most (maybe all?) cases I found. No sorting, would be lightning fast.

File: utils/luafunc_dup.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/bash
+
+LST=`tempfile`
```
Comment (bobbens) [inline utils/luafunc_dup.sh:3]:
```suggestion
LST=$(mktemp)
```
Apparently `tempfile` is deprecated and backtick notation is considered legacy.

File: utils/luafunc_dup.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+res=0
+trap 'rm -f $LST ; exit $res' EXIT
+
+for arg in $@ ; do
```
Comment (bobbens) [inline utils/luafunc_dup.sh:7]:
```suggestion
for arg in "$@ "; do
```
Avoid re-splitting elements.

File: utils/luafunc_dup.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+   grep '^ \* @luafunc' $arg | sort | uniq -c | grep -v "^ *1 " > $LST
+
+   if [ -s "$LST" ] ; then
+      echo "$arg:"
+      cat $LST
```
Comment (bobbens) [inline utils/luafunc_dup.sh:12]:
```suggestion
      cat "$LST"
```
More potential resplitting.

File: utils/luafunc_dup.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/bash
+
+LST=`tempfile`
+res=0
+trap 'rm -f $LST ; exit $res' EXIT
+
+for arg in $@ ; do
+   grep '^ \* @luafunc' $arg | sort | uniq -c | grep -v "^ *1 " > $LST
```
Comment (bobbens) [inline utils/luafunc_dup.sh:8]:
```suggestion
   grep '^ \* @luafunc' "$arg" | sort | uniq -c | grep -v "^ *1 " > "$LST"
```

**Review by bobbens — CHANGES_REQUESTED** [review]
Some changes point out by pre-commit shellcheck.

File: utils/luafunc_dup.sh (hunk 2 above)
Comment (johann1764) [reply utils/luafunc_dup.sh:7]: Absolutely !
That said, I think your `'"'` and `' '` were swapped.

File: utils/luafunc_dup.sh (hunk 3 above)
Comment (johann1764) [reply utils/luafunc_dup.sh:12]: LST is a tempfile name, so shouldn't contain spaces.
But better safe than sorry, let us do this.

File: utils/luafunc_dup.sh (hunk 4 above)
Comment (johann1764) [reply utils/luafunc_dup.sh:8]: True.

File: utils/luafunc_dup.sh (hunk 2 above)
Comment (bobbens) [reply utils/luafunc_dup.sh:7]: Thanks for the notice, I dislike the github editor.

File: utils/luafunc_dup.sh (hunk 1 above)
Comment (johann1764) [reply utils/luafunc_dup.sh:3]: Damnit, I am feeling old, now.
But true, someone already told me that backtick notation was not used anymore, a long time ago.

**Comment by bobbens** [comment]
Looks good, thanks! Seems to have caught a lot of them already.

**Comment by johann1764** [comment]
Should be all of them. (I only left the one you already corrected in main)
